src/core/utils/fs.cpp

```diff
@@ -112,57 +113,51 @@ QString Utils::Fs::folderName(const QString& file_path)
 }
 
 /**
- * Remove an empty folder tree.
- *
- * This function will also remove .DS_Store files on Mac OS and
- * Thumbs.db on Windows.
+ * This function will first remove system cache files, e.g. `Thumbs.db`,
+ * `.DS_Store`. Then will try to remove the whole tree if the tree consist
+ * only of folders
  */
-bool Utils::Fs::smartRemoveEmptyFolderTree(const QString& dir_path)
+bool Utils::Fs::smartRemoveEmptyFolderTree(const QString& path)
 {
-    qDebug() << Q_FUNC_INFO << dir_path;
-    if (dir_path.isEmpty())
+    if (!QDir(path).exists())
         return false;
 
-    QDir dir(dir_path);
-    if (!dir.exists())
-        return true;
-
-    // Remove Files created by the OS
-#if defined Q_OS_MAC
-    forceRemove(dir_path + QLatin1String("/.DS_Store"));
-#elif defined Q_OS_WIN
-    forceRemove(dir_path + QLatin1String("/Thumbs.db"));
+    static const QStringList deleteList = {
+#if defined Q_OS_WIN
+    "Thumbs.db", "desktop.ini"
+#elif defined Q_OS_MAC
+    ".DS_Store"
 #endif
-
-    QFileInfoList sub_files = dir.entryInfoList();
-    foreach (const QFileInfo& info, sub_files) {
-        QString sub_name = info.fileName();
-        if (sub_name == "." || sub_name == "..")
-            continue;
-
-        QString sub_path = info.absoluteFilePath();
-        qDebug() << Q_FUNC_INFO << "sub file: " << sub_path;
-        if (info.isDir()) {
-            if (!smartRemoveEmptyFolderTree(sub_path)) {
-                qWarning() << Q_FUNC_INFO << "Failed to remove folder: " << sub_path;
-                return false;
-            }
+    };
+
+    // travel from the deepest folder and remove anything unwanted on the way out.
+    QStringList dirList(path + "/");  // get all sub directories paths
+    QDirIterator iter(path, (QDir::AllDirs | QDir::NoDotAndDotDot), QDirIterator::Subdirectories);
+    while (iter.hasNext())
+        dirList << iter.next() + "/";
+    std::sort(dirList.begin(), dirList.end(), [](const QString &l, const QString &r){return (l.count("/") > r.count("/"));});  // sort descending by directory depth
+
+    for(const QString &p : dirList) {
```

Comment:
I'm a little confused... I found the recent tests. For Qt containers, all methods do not differ in performance. But for std containers Qt foreach contraindicated because it does a deep copy (easy for Qt containers, because they implement the copy-on-write strategy).
So sorry that I forced you to fix it. c++11 for(each) is preferable, because it gives an universal approach to all container types.